linkml/generators/rustgen/rustgen.py

```diff
@@ -178,55 +210,55 @@ def get_rust_type(t: Union[TypeDefinition, type, str], sv: SchemaView, pyo3: boo
+            no_add_crate = True
         elif t in sv.all_classes():
             c = sv.get_class(t)
             rsrange = get_name(c)
             descendants = sv.class_descendants(t)
-            if len(descendants) > 1:
-                rsrange += "OrSubtype"
+            #if len(descendants) > 1:
+            #    rsrange += "OrSubtype"
 
     # FIXME: Raise here once we have implemented all base types
     if rsrange is None:
         rsrange = PYTHON_TO_RUST[str]
     elif rsrange in PYTHON_TO_RUST:
         rsrange = PYTHON_TO_RUST[rsrange]
+    elif crate_ref is not None and not no_add_crate:
+        rsrange = f"{crate_ref}::{rsrange}"
     return rsrange
 
 
-def get_rust_range(cls: ClassDefinition, s: SlotDefinition, sv: SchemaView, range: Optional[str] = None) -> str:
-    range = range if range is not None else s.range
-    boxing_needed = False
-    inline_mode = get_inline_mode(s, sv)
-    base_type = get_rust_type(range, sv, True)
-    all_ranges = sv.slot_range_as_union(s)
-    if len(all_ranges) > 1:
-        base_type = underscore(uncamelcase(cls.name)) + "_utl::" + get_name(s) + "_range"
-
-
-    if inline_mode == SlotInlineMode.IDENTIFIER:
-        base_type = "String"
-    else:
-        if can_contain_reference_to_class(s, cls, sv):
-            boxing_needed = True
 
-    if boxing_needed:
-        base_type = f"Box<{base_type}>"
+def get_rust_range_info(cls: ClassDefinition, s: SlotDefinition, sv: SchemaView, crate_ref: Optional[str] = None) -> RustRange:
+    (container_mode, inline_mode) = get_inline_mode(s, sv)
+    all_ranges = sv.slot_range_as_union(s)
+    sub_ranges = [
+        RustRange(type_="String" if inline_mode == SlotInlineMode.REFERENCE else get_rust_type(r, sv, True, crate_ref), 
+                  is_class_range = r in sv.all_classes(),
+                  has_class_subtypes= len(sv.class_descendants(r)) > 1 if r in sv.all_classes() else False,
+                  )
+        for r in all_ranges
+    ]
+    
+    res = RustRange(
+        optional=not s.required,
+        has_default = not (s.required or False) or (s.multivalued or False),
+        containerType= ContainerType.LIST if container_mode == SlotContainerMode.LIST else ContainerType.MAPPING if container_mode == SlotContainerMode.MAPPING else None,
+        child_ranges=sub_ranges if len(sub_ranges) > 1 else None,
+        box_needed=inline_mode == SlotInlineMode.INLINE and can_contain_reference_to_class(s, cls, sv),
+        is_class_range=all_ranges[0] in sv.all_classes() if len(all_ranges) == 1 else False,
+        is_reference=inline_mode == SlotInlineMode.REFERENCE,
+        has_class_subtypes= len(sv.class_descendants(all_ranges[0])) > 1 if len(all_ranges) == 1 and all_ranges[0] in sv.all_classes() else False,
+        type_ = underscore(uncamelcase(cls.name)) + "_utl::" + get_name(s) + "_range" if len(sub_ranges) > 1 else ("String" if inline_mode == SlotInlineMode.REFERENCE else get_rust_type(s.range, sv, True, crate_ref))
+    )
```

Comment:
maybe the most meaty set of one-liners that all make sense in context that i've ever seen in python